snuba/pipeline/composite_entity_processing.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import annotations
+
+from typing import Mapping, Union
+
+from snuba.clickhouse.query import Query as ClickhouseQuery
+from snuba.datasets.plans.entity_processing import run_entity_processing_executor
+from snuba.pipeline.composite_storage_processing import check_sub_query_storage_sets
+from snuba.query import ProcessableQuery
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.join import IndividualNode, JoinClause, JoinVisitor
+from snuba.query.data_source.simple import Entity, Table
+from snuba.query.data_source.visitor import DataSourceVisitor
+from snuba.query.joins.equivalence_adder import add_equivalent_conditions
+from snuba.query.joins.subquery_generator import generate_subqueries
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.query_settings import QuerySettings
+
+
+def translate_composite_query(
+    query: CompositeQuery[Entity], query_settings: QuerySettings
+) -> CompositeQuery[Table]:
+    """
+    A function that traverses a logical composite query, applies all entity processors,
+    and translates the result into a physical composite query.
+    """
+    add_equivalent_conditions(query)
+    generate_subqueries(query)
```

Comment:
We generate subqueries from join query and push down all expressions that can be executed in those subquery because it helps the pipeline more easily apply storage processors to each node. When subquery node references an individual storage; the pipeline simply just needs to traverse through each node, and apply their respective processors for that datasource.